planner/core/exhaust_physical_plans.go

```diff
@@ -2351,23 +2342,23 @@ func (la *LogicalAggregation) getHashAggs(prop *property.PhysicalProperty) []Phy
 		taskTypes = append(taskTypes, property.CopTiFlashLocalReadTaskType)
 	}
 	canPushDownToMPP := la.ctx.GetSessionVars().AllowMPPExecution && !collate.NewCollationEnabled() && la.checkCanPushDownToMPP()
-	if canPushDownToMPP {
-		taskTypes = append(taskTypes, property.MppTaskType)
-	}
 	if la.HasDistinct() {
 		// TODO: remove after the cost estimation of distinct pushdown is implemented.
-		if !la.ctx.GetSessionVars().AllowDistinctAggPushDown && !canPushDownToMPP {
+		if !la.ctx.GetSessionVars().AllowDistinctAggPushDown {
 			taskTypes = []property.TaskType{property.RootTaskType}
 		}
 	} else if !la.aggHints.preferAggToCop {
 		taskTypes = append(taskTypes, property.RootTaskType)
 	}
+	if !la.canPushToCop() {
+		taskTypes = []property.TaskType{property.RootTaskType, property.CopTiFlashLocalReadTaskType}
+	}
+	if canPushDownToMPP {
+		taskTypes = append(taskTypes, property.MppTaskType)
+	}
 	if prop.IsFlashProp() {
 		taskTypes = []property.TaskType{prop.TaskTp}
 	}
```

Comment:
the upper code should be in a else ?

Reply:
no. being able to push down to cop doesn't mean not to push down to mpp.